### PR TITLE
 Adds support for coordinate subscripting.

### LINKED ISSUE #366: Make coordinate objects subscriptable
🐞 **Problem**

We can't do a[0] with the coordinates right now, which is very intuitive. Thanks to @shilpi1958 , we know what needs to be done!
🎯 **Goal**

💡 **Possible solutions**

📋  **Steps to solve the problem**

 * Comment below about what you've started working on.
 * Add, commit, push your changes
 * Submit a pull request and add this in comments - `Addresses #<put issue number here>`
 * Ask for a review in comments section of pull request
 * Celebrate your contribution to this project 🎉

### PR DESCRIPTION
**Pull request by Raahul-Singh**
Fixes #366

### CODE REVIEW

**Comment by shreyasbapat** [comment]
Hey @Raahul-Singh you need to take care of code linting. 

You will have to run `tox -e reformat` to fix this :)

**Comment by Raahul-Singh** [comment]
> Hey @Raahul-Singh you need to take care of code linting.
> 
> You will have to run `tox -e reformat` to fix this :)

And with one command you have helped me solve most of my pep8 issues. Thanks for this!!

File: src/einsteinpy/coordinates/core.py [hunk 1]
```diff
@@ -30,13 +30,22 @@ def __init__(self, x, y, z):
         self.z = z
         super().__init__(x.si.value, y.si.value, z.si.value)
         self.system = "Cartesian"
+        self.coordinates = {
```
Comment (shreyasbapat) [inline src/einsteinpy/coordinates/core.py]: I guess the self.coordinates is a vague way of indicating what's already coordinate!

File: src/einsteinpy/coordinates/core.py [hunk 2]
```diff
@@ -30,13 +30,22 @@ def __init__(self, x, y, z):
...
+        }
 
     def __repr__(self):
         return "Cartesian x: {}, y: {}, z: {}".format(self.x, self.y, self.z)
 
     def __str__(self):
         return self.__repr__()
 
+    def __getitem__(self, item):
+        return self.coordinates[item]
```
Comment (shreyasbapat) [inline src/einsteinpy/coordinates/core.py]: Same here

**Comment by shreyasbapat** [comment]
Also, this would need a rebase

File: src/einsteinpy/coordinates/core.py (hunk 1 above)
Comment (Raahul-Singh) [reply src/einsteinpy/coordinates/core.py]: Should I go with `self.data` ?

**Comment by Raahul-Singh** [comment]
> Can we also write unit tests for this, that would make the PR mergeable, along with the changes O proposed.
> Can we do `self._dimension` instead ?

Where should I put the test? Which file?
I'm a little confused as this is a simple `__getitem__` implementation.

**Comment by shreyasbapat** [comment]
Sure. You can put tests in this file: 

https://github.com/einsteinpy/einsteinpy/blob/master/src/einsteinpy/tests/test_coordinates/test_coord_transform.py

**Comment by shreyasbapat** [comment]
Still, we need to be sure that it works always! :)

**Comment by ritzvik** [comment]
My two cents for this PR.
Can we make `obj[0]`, `obj[1]` and `obj[2]` the aliases for `obj["x"]`, `obj["y"]` and `obj["z"]` ? Same for `r, theta, phi`. 

Anyway, @Raahul-Singh , wait for @shreyasbapat go-ahead before implementing this.

**Comment by shreyasbapat** [comment]
My end goal with the issue was the same. But then it came to my mind, how do we know which one is 0,1 and 2. If you feel this is not a big issue @ritzvik , and have some solution. I am more inclined to the solution you suggested.

**Comment by ritzvik** [comment]
I guess it's pretty evident in Cartesian case, `0,1,2` for `x,y,z`.
`r, theta, phi` is the order used in [wikipedia](https://en.wikipedia.org/wiki/Spherical_coordinate_system), so we can go with that in case of Spherical and BL.

We can add these entries into the same `self.coordinates` as coded by @Raahul-Singh , where `0,1,2` can be the keys and `self.x, self.y, self.z` can be the values. What do you say @shreyasbapat ?

Also, @Raahul-Singh , please change the same to another other than `self.coordinates`. It seems very ambiguous

**Comment by shreyasbapat** [comment]
If you feel that's obvious, I am fine. 
Rename it to `self._dimensions`

**Comment by shreyasbapat** [comment]
This can't show up in the docs. We need to make it private

**Comment by Raahul-Singh** [comment]
Hey! I've changed `self.coordinates` to `self._dimension` . I've also added a test. The simplest way to make this subscript-able by index is to use a tuple rather than a dictionary. Tuple instead of a list because list is mutable. However, `self._dimension[-1]` would get the name of the coordinate whereas all others would get quantities. Would this not cause occasional errors, if the user is not familiar with it?

File: src/einsteinpy/coordinates/core.py [hunk 3]
```diff
@@ -30,13 +30,22 @@ def __init__(self, x, y, z):
         self.z = z
         super().__init__(x.si.value, y.si.value, z.si.value)
         self.system = "Cartesian"
+        self._dimension = {
+            "x": self.x,
+            "y": self.y,
+            "z": self.z,
+            "system": "Cartesian",
+        }
```
Comment (ritzvik) [inline src/einsteinpy/coordinates/core.py]: Add a line here,
```suggestion
        self._dimension_order = ("x", "y", "z")
```

File: src/einsteinpy/coordinates/core.py [hunk 4]
```diff
@@ -30,13 +30,22 @@ def __init__(self, x, y, z):
...
 
     def __repr__(self):
         return "Cartesian x: {}, y: {}, z: {}".format(self.x, self.y, self.z)
 
     def __str__(self):
         return self.__repr__()
 
+    def __getitem__(self, item):
+        return self._dimension[item]
```
Comment (ritzvik) [inline src/einsteinpy/coordinates/core.py]: And a change here,
```suggestion
        if isinstance(item, (int, np.integer)):
            return self._dimension[self._dimension_order[item]]
        return self._dimension[item]
```

Does this two comments address this https://github.com/einsteinpy/einsteinpy/pull/387#issuecomment-578191030 ?

Comment (ritzvik) [reply src/einsteinpy/coordinates/core.py]: Also, please write a docstring for `__getitem__`s. Users should know that objects are subscriptable.

**Comment by shreyasbapat** [comment]
> Hey! I've changed `self.coordinates` to `self._dimension` . I've also added a test. The simplest way to make this subscript-able by index is to use a tuple rather than a dictionary. Tuple instead of a list because list is mutable. However, `self._dimension[-1]` would get the name of the coordinate whereas all others would get quantities. Would this not cause occasional errors, if the user is not familiar with it?

I guess we will document that somewhere. Put a comment in the code for the same. I will get it sorted in my next PR. And please fix the sggestions by @ritzvik . And rebase and squash!

**Comment by shreyasbapat** [comment]
@Raahul-Singh are you still on it?

**Comment by Raahul-Singh** [comment]
> @Raahul-Singh are you still on it?

Oh, yes. I'll get it done by tonight. Just a busy couple of days. 😅 I'm sorry.

**Comment by Raahul-Singh** [comment]
Does this work?

**Comment by ritzvik** [comment]
> Does this work?

Yups! Seems good enough!
BTW, maybe, you rebased the branch wrong. The PR is showing "25 files changed".

**Comment by Raahul-Singh** [comment]
> > Does this work?
> 
> Yups! Seems good enough!
> BTW, maybe, you rebased the branch wrong. The PR is showing "25 files changed".

Sorry. :sweat_smile: I am not that experienced with rebasing yet. Could you help me with fixing it?

File: src/einsteinpy/coordinates/core.py [hunk 5]
```diff
@@ -217,6 +263,14 @@ def __init__(self, r, theta, phi, a):
...
         super().__init__(r.si.value, theta.si.value, phi.si.value, a=a.si.value)
         self.system = "BoyerLindquist"
+        self._dimension = {
+            "r": self.r,
+            "theta": self.theta,
+            "phi": self.phi,
+            "a": self.a,
+            "system": "BoyerLindquist",
+        }
+        self._dimension_order = ("r", "theta", "phi", "a")
```
Comment (ritzvik) [inline src/einsteinpy/coordinates/core.py]: "a" is not a dimension here. It is an attribute of space-time. Can you remove it from `_dimension_order`?

Comment (ritzvik) [reply src/einsteinpy/coordinates/core.py]: P.S. : feature of the space spanned by BL coordinates

Comment (ritzvik) [reply src/einsteinpy/coordinates/core.py]: This one thing, and then we can rebase!

**Comment by ritzvik** [comment]
I guess you somehow included all the previous commits in 
21d9a97 .

So I would say, go to the project root and do this:
```
git reset --hard HEAD~3 # makes a hard reset to 3 commits behind
git checkout dfe2eac -- ./src/einsteinpy/coordinates/core.py # copies the file from your latest commit
```
You would find `core.py` file staged and make a new commit. After this, rebase with upstream
```
git pull --rebase upstream master
```

I believe the changes in test file are pretty small. Just copy-paste them and create another commit.

That's it

**Comment by Raahul-Singh** [comment]
Thanks @ritzvik :smile:

**Comment by ritzvik** [comment]
@shreyasbapat LGTM.
If you are satisfied, merge it.

File: src/einsteinpy/coordinates/core.py [hunk 6]
```diff
@@ -30,13 +30,36 @@ def __init__(self, x, y, z):
...
+        Objects are subsctiptable with both explicit names of parameters
+        and integer indices.
+
+        Parameters
+        ----------
+        item : ~astropy.units.quantity.Quantity or integer
+                Name of the parameter or it's index.
+                If `system`, Name of coordinate is returned.
+        """
+        if isinstance(item, (int, np.integer)):
```
Comment (shreyasbapat) [inline src/einsteinpy/coordinates/core.py]: Very good @Raahul-Singh ! This is exactlt what was expected!

Comment (Raahul-Singh) [reply src/einsteinpy/coordinates/core.py]: Thank you for helping me along the way!

**Pull request merged**